### ISSUE TITLE
incorrect focal loss implementation mathematically neutralizing weighting effect

### ISSUE THREAD

**Issue by 0xzcr** [foamliu/InsightFace-v2#38]
**The Issue**
right now, the current implementation of `FocalLoss` in `focal_loss.py` initializes `nn.CrossEntropyLoss` with default settings. In PyTorch, the default `reduction` for `CrossEntropyLoss` is `'mean'`. 
this causes the loss to be reduced to a single scalar before the focal weights are applied. as a result, the focal weighing is applied to the global average loss rather than to individual hard/easy samples, which completely defeats the purpose of Focal Loss (hard-example mining).

**Affected Code**
file - focal_loss.py

```python
10:         self.ce = torch.nn.CrossEntropyLoss()
11: 
12:     def forward(self, input, target):
13:         logp = self.ce(input, target)
14:         p = torch.exp(-logp)
15:         loss = (1 - p) ** self.gamma * logp
16:         return loss.mean()
```
**Impact**
the model learns from all samples with equal priority, meaning hard examples do not recieve the intended gradient boost, and easy examples are not down-weighted. This leads to suboptimal convergence on imbalanced or difficult datasets.

**The fix**

CrossEntropyLoss should be initialized with `reduction='none'` to return per-sample losses. 
the focal weights should then be applied to these individual losses before computing the final mean.

**what i propose in code**
**_focal_loss.py_**
```python
class FocalLoss(nn.Module):
    def __init__(self, gamma=0):
        super(FocalLoss, self).__init__()
        self.gamma = gamma
        self.ce = torch.nn.CrossEntropyLoss(reduction='none') # <---putting the change here

    def forward(self, input, target):
        logp = self.ce(input, target) # This is now a vector [N]
        p = torch.exp(-logp)
        loss = (1 - p) ** self.gamma * logp
        return loss.mean()
```

**Pre and Post Test Results**
i ran a test case with two samples:
*   Sample 0 (Hard): The model is completely wrong (Cross-Entropy = 15.0).
*   Sample 1 (Easy):  The model is perfectly right (Cross-Entropy = 0.0).

(current logic) :- 
Logic: Calculate mean CE (7.5000) -> Apply one weight based on mean
Global Weight applied: 0.9989
Final Scaler Loss: 7.491707

where the problem lies - the signal from the hard sample (15.0) was averaged with the easy sample (0.0) the model could identify which was which. the model thinks both samples are "moderately difficult" (7.5) and assigns them the same weight.

(fixed logic) :- 
in my solution, weights are applied to each sample individually.
Logic: Apply weights per sample -> Calculate mean of weighted results
Sample 0 Weighted Loss: 15.0000 (Weight: 1.0000)
Sample 1 Weighted Loss: 0.0000 (Weight: 0.0000)
Final Scaler Loss: 7.499996

i hope my findings are relevant and help contribute to improving the model. :) 
thank you
